HeterogeneousCore/CUDAUtilities/src/CUDAStreamCache.cc

```diff
@@ -0,0 +1,42 @@
+#include "HeterogeneousCore/CUDAUtilities/interface/CUDAStreamCache.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/currentDevice.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/ScopedSetDevice.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaDeviceCount.h"
+
+namespace cudautils {
+  void CUDAStreamCache::Deleter::operator()(cudaStream_t stream) const {
+    if (device_ != -1) {
+      ScopedSetDevice deviceGuard{device_};
+      cudaCheck(cudaStreamDestroy(stream));
+    }
+  }
+
+  // CUDAStreamCache should be constructed by the first call to
+  // getCUDAStreamCache() only if we have CUDA devices present
+  CUDAStreamCache::CUDAStreamCache() : cache_(cudautils::cudaDeviceCount()) {}
+
+  SharedStreamPtr CUDAStreamCache::getCUDAStream() {
+    const auto dev = cudautils::currentDevice();
+    return cache_[dev].makeOrGet([dev]() {
+      cudaStream_t stream;
+      cudaCheck(cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking));
+      return std::unique_ptr<BareStream, Deleter>(stream, Deleter{dev});
+    });
+  }
+
+  void CUDAStreamCache::clear() {
+    // Reset the contents of the caches, but leave an
+    // edm::ReusableObjectHolder alive for each device. This is needed
+    // mostly for the unit tests, where the function-static
+    // CUDAStreamCache lives through multiple tests (and go through
+    // multiple shutdowns of the framework).
+    cache_.clear();
+    cache_.resize(cudautils::cudaDeviceCount());
+  }
+
+  CUDAStreamCache& getCUDAStreamCache() {
+    static CUDAStreamCache cache;
```

Comment:
Static analyzer complains about non-const static variable. The public interface of the object is thread safe, so `CMS_SA_ALLOW` would be the proper way to go.